src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -1032,6 +1014,81 @@ private static string CreateFileDictionaryKey(FileData v2File, IList<FileData> v
             return logicalLocationsVersionOne;
         }
 
+        // In SARIF v1, run.resources.rules was a dictionary, whereas in v2 it is an array.
+        // Normally, the lookup key into the v1 rules dictionary is the rule id. But some
+        // tools allow multiple rules to have the same id. In that case we must synthesize
+        // a unique key for each rule with that id. We choose "<ruleId>-<n>", where <n> is
+        // 1 for the second occurrence, 2 for the third, and so on.
+        private static IDictionary<int, string> CreateV2RuleIndexToV1KeyMapping(IList<Rule> rules)
+        {
+            var v2RuleIndexToV1KeyMap = new Dictionary<int, string>();
+
+            if (rules != null)
+            {
+                // Keep track of how many distinct rules have each id.
+                var ruleIdToCountMap = new Dictionary<string, int>();
+
+                for (int i = 0; i < rules.Count; ++i)
+                {
+                    Rule rule = rules[i];
+                    if (rule.Id != null)
```

Comment:
>if (rule.Id != null) [](start = 20, length = 20)

ok, so in old SARIF, if you had no rule.Id, you had no look-up key, so there won't be any coupling between results and rule metadata for these checks.